### PR TITLE
Add end-to-end tests for `iree_linalg_ext.custom_op`.

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
This PR plumbs through e2e tests for `iree_linalg_ext.custom_op` on
CPU and AMDGPU backends. The end-to-end test does a matmul + bias add
with and without the use of `custom_op` and checks that the same value
is obtained.

Signed-off-by: MaheshRavishankar <mahesh.ravishankar@gmail.com>

### CODE REVIEW

**Comment by MaheshRavishankar** [comment]
Needs upstream patch to submit this, but apart from that this is ready for review

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerExecutableTarget.cpp [hunk 1]
```diff
@@ -127,37 +130,57 @@ void LLVMCPULowerExecutableTargetPass::runOnOperation() {
     break;
   case IREE::Codegen::DispatchLoweringPassPipeline::
       CPUBufferOpsTileAndVectorize: {
-    TilingConfig tilingConfig = getTilingConfigForPipeline(funcOp);
-    addCPUBufferOpsTileAndVectorizePipeline(pipeline, tilingConfig,
+    FailureOr<TilingConfig> tilingConfig = getTilingConfigForPipeline(funcOp);
+    if (failed(tilingConfig)) {
+      return signalPassFailure();
+    }
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerExecutableTarget.cpp]: This code is duplicated for all cases except default and none. Maybe split into two `switch` statements?

File: compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/DecomposeAggregateOps.cpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
+// Copyright 2024 The IREE Authors
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/DecomposeAggregateOps.cpp:1]: Can you add a test for this pass with a positive test for `custom_op` and a negative test for another aggregate op (maybe linalg.softmax)

File: tests/e2e/linalg_ext_ops/BUILD.bazel [hunk 3]
```diff
@@ -38,6 +41,21 @@ iree_check_single_backend_test_suite(
     target_backend = "llvm-cpu",
 )
 
+iree_check_single_backend_test_suite(
+    name = "check_no_dt_llvm-cpu_local-task",
+    srcs = [
+        "custom_op.mlir",
+    ],
+    compiler_flags = [
+        "--iree-opt-data-tiling=false",
```
Comment (qedawkins) [inline tests/e2e/linalg_ext_ops/BUILD.bazel:71]: Is this separated because data tiling applies to the contents of the custom op? We should [add a TODO to] change that if so.

**Review by qedawkins — APPROVED** [review]

File: compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/DecomposeAggregateOps.cpp [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  // Use the initializer to set default options for ListOption.
+  LogicalResult initialize(MLIRContext *context) override {
+    this->decomposeOps = {IREE::LinalgExt::CustomOp::getOperationName().str()};
+    return success();
+  }
+
+  void runOnOperation() override;
+};
+
+void DecomposeAggregateOpsPass::runOnOperation() {
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/DecomposeAggregateOps.cpp:32]: optional style nit: I know that the coding style is not consistent in IREE codebase, but I'd like to try hard to make it happen. Can you move the implementation out of anonymous namespaces?

Inlining the reason from LLVM style guide, and I personally like the idea:

> The problem with anonymous namespaces is that they naturally want to encourage indentation of their body, and they reduce locality of reference: if you see a random function definition in a C++ file, it is easy to see if it is marked static, but seeing if it is in an anonymous namespace requires scanning a big chunk of the file.
>
> Because of this, we have a simple guideline: make anonymous namespaces as small as possible, and only use them for class declarations.

https://llvm.org/docs/CodingStandards.html#anonymous-namespaces

File: tests/e2e/linalg_ext_ops/BUILD.bazel (hunk 3 above)
Comment (hanhanW) [reply tests/e2e/linalg_ext_ops/BUILD.bazel:71]: Yeah, I think so. We should teach SetEncoding about it. The SetEncoding pass skips the ops with pre-set configuration, and I think we need to teach it about CustomOp as well.

https://github.com/iree-org/iree/blob/103d631f8ba53954eccac821a17e6d2a15918b32/compiler/src/iree/compiler/DispatchCreation/SetEncoding.cpp#L169-L172

**Comment by benvanik** [comment]
tests are always good; @MaheshRavishankar, is this something you can rebase or is it stale?

**Comment by MaheshRavishankar** [comment]
Its on my back log to rebase and land this. Have been having unrelated failures.
